include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include "rmm/cuda_stream_view.hpp"
+
+#include <cuda_runtime_api.h>
+
+#if CUDART_VERSION >= 11020  // 11.2 introduced cudaMallocAsync
+#define CUDA_MALLOC_ASYNC_SUPPORT
+#endif
+
+namespace rmm {
+namespace mr {
+
+/**
+ * @brief `device_memory_resource` derived class that uses `cudaMallocAsync`/`cudaFreeAsync` for
+ * allocation/deallocation.
+ */
+class cuda_async_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Default constructor
+   *
+   * @throws rmm::runtime_error if the CUDA version does not support `cudaMallocAsync`
+   */
+  cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    // Check if cudaMallocAsync Memory pool supported
+    int device{0};
+    RMM_CUDA_TRY(cudaGetDevice(&device));
+    int v{0};
+    auto e = cudaDeviceGetAttribute(&v, cudaDevAttrMemoryPoolsSupported, device);
+    RMM_EXPECTS(e == cudaSuccess && v == 1,
+                "cudaMallocAsync not supported with this CUDA driver/runtime version");
+#else
+    RMM_FAIL("cudaMallocAsync not supported");
+#endif
+  }
+
+  ~cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    cudaDeviceSynchronize();
+    int device{0};
+    RMM_ASSERT_CUDA_SUCCESS(cudaGetDevice(&device));
+    cudaMemPool_t pool;
+    RMM_ASSERT_CUDA_SUCCESS(cudaDeviceGetDefaultMemPool(&pool, device));
+    RMM_ASSERT_CUDA_SUCCESS(cudaMemPoolTrimTo(pool, 0));
```

Comment:
Question: I though we should first trim and then synchronize?

Reply:
Yeah, I'm new to cudaMallocAsync too. I wasn't exactly sure what I should do...  Do you have a reference to where you got that info?

Reply:
I got this impression from a bug candidate I found (see https://github.com/cupy/cupy/issues/3777#issuecomment-758842843 and @jrhemstad's reply right below). I think the issue there was `cudaMemPoolTrimTo` does not release memory immediately, so if I let the device mempool to hold almost all memory and then do `freeAsync` -> `trim`, any subsequent `cudaMalloc` would fail unless I sync (over something, be it stream or device) . 

I don't have access to my CUDA 11.2 machine ftm, though, so I can't verify if a reverse order is ok or not.

Reply:
I expect all memory allocated using this MR to be freed using this MR *before* the MR destructor is called. Therefore, I would expect that synchronizing the device first would be basically the same as synchronizing after trimming.

Reply:
Strictly speaking, I don't think trimming the pool is necessary. `cudaMalloc` will automatically invoke garbage collection if it attempts to allocate memory and fails.

Reply:
> I expect all memory allocated using this MR to be freed using this MR _before_ the MR destructor is called.

That's right, not sure what I was thinking last night 😅 

> Strictly speaking, I don't think trimming the pool is necessary. `cudaMalloc` will automatically invoke garbage collection if it attempts to allocate memory and fails.

I am less certain about the gc part, but I agree it's not necessary to trim the pool, especially if it's the default pool which in theory could still be used by other applications at the time of MR destruction (if I understand RMM correctly...)

Reply:
Agree we should not trim the default pool.